### PR TITLE
add art libraries to manage Git repos, RPMs and versions in Jenkins jobs

### PR DESCRIPTION
**Pull request by markllama**
This change adds a set of libraries that can be used in accord with the standard Groovy/Jenkins library mechanisms.  This avoids the single file require clauses and unifies the library structures.

These libraries are object-oriented and allow the treatment of the components as types and classes. For example it allows comparison of version strings in an idiomatic way in Groovy, simplifying the code and offering clearer interpretation by the reader.

### CODE REVIEW

**Comment by markllama** [comment]
This change has no conflict with existing code.  It does not change anything existing and is not yet used by any jobs.  It is being submitted in preparation for the presentation of new jobs that will decompose the ocp build job so that the build pipeline can be optimized and so retries of individual components will not require complete rebuilds.

File: src/com/redhat/art/VersionTest.groovy [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import com.redhat.art.Version
+
+class VersionTest extends GroovyTestCase {
+    def env
+    
+    def VersionTest(pipeline_env=null) {
+        env = pipeline_env
+    }
+
+    //protected Version v0
```
Comment (adammhaile) [inline src/com/redhat/art/VersionTest.groovy]: Needed anymore? Clean up commented stuff.

File: src/com/redhat/art/VersionTest.groovy [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        assertEquals(actual, expected)
+    }
+
+    void testCompareTo() {
+
+        def inputs = [
+            [
+                testValue: new Version('3.2.1'),
+                greater: [
+                    new Version('3.2.1.1'),
```
Comment (adammhaile) [inline src/com/redhat/art/VersionTest.groovy]: What's all this?

File: src/com/redhat/art/RpmSpecTest.groovy [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    String samples
+
+    void setUp() {
+        samples = [
+            '''# Filler
+Version: 3.0
+Release: 1
+%changelog
+#
+Here is some stuff
```
Comment (adammhaile) [inline src/com/redhat/art/RpmSpecTest.groovy]: Stuff? This is not final text is it?

File: src/com/redhat/art/RpmSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    def pipeline
+
+    def RpmSpec(Map init) {
+        this.filename = init.filename
+        this.lines = init.lines
+        this.pipeline = init.pipeline
+
+    }
+    
+    //def RpmSpec(content) {
```
Comment (adammhaile) [inline src/com/redhat/art/RpmSpec.groovy]: Remove

File: src/com/redhat/art/RpmIdentifierTest.groovy [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.redhat.art
```
Comment (adammhaile) [inline src/com/redhat/art/RpmIdentifierTest.groovy]: What's the point of this file?

File: src/com/redhat/art/RpmIdentifier.groovy [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    def RpmIdentifier(Version v, Version r) {
+        version = v
+        release = r
+    }
+
+    def RpmIdentifier(string v_string, string r_string) {
+        version = new Version(v_string)
+        release = new Version(r_string)
+    }
+
```
Comment (adammhaile) [inline src/com/redhat/art/RpmIdentifier.groovy]: Extra new lines

File: src/com/redhat/art/GitHubRepository.groovy [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                "git clone",
+                branch ? "--branch ${branch}" : "",
+                remote,
+                path
+            ].join(' ')
+        )
+        pipeline.echo("Cloning repo ${remote}")
+    }
+    
+    /*
```
Comment (adammhaile) [inline src/com/redhat/art/GitHubRepository.groovy]: Delete

File: src/com/redhat/art/VersionTest.groovy (hunk 2 above)
Comment (markllama) [reply src/com/redhat/art/VersionTest.groovy]: All of the values in the list should test "greater" than 3.2.1.  3.2.1 is implicit 3.2.1.0  So 3.2.1.1 should test greater.

File: src/com/redhat/art/RpmSpecTest.groovy (hunk 3 above)
Comment (markllama) [reply src/com/redhat/art/RpmSpecTest.groovy]: Actually that's perfectly good final text. This is a mock, not something that will actually commit
 Now I do actually have to *write the tests*...

Comment (markllama) [reply src/com/redhat/art/RpmSpecTest.groovy]: Removed null test file

**Comment by markllama** [comment]
@adammhaile  time for another review:
I removed unused/duplicate libraries (RpmIdentifier == Version) and empty tests.

**Pull request merged**